design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+    kind: AggregateResourceClass
+    name: cool-gke-cluster
+    uid: ea414c74-0cd9-11ea-8d71-362b9e155667
+  writeConnectionSecretsToRef:
+    namespace: crossplane-system
+    name: eabce854-0cd7-11ea-8d71-362b9e155667
+```
+
+The purpose of these new resources, as their names imply, is to _aggregate_
+resource classes and managed resources, enabling transitive one-to-many
+resource-claim-to-resource-class and resource-claim-to-managed-resource
+relationships. Both new kinds use an _aggregation rule_, like an [RBAC
+`ClusterRole`]. The `AggregateResourceClass` aggregates resource classes (of any
+non-aggregate kind) by label, while the `AggregateResource` aggregates managed
+resources (again, of any non-aggregate kind) by label.
+
+In the above example `default-coolcluster-g3bf7` is an `AggregateResource` that
+was dynamically provisioned to satisfy a `KubernetesCluster` resource claim. An
+`AggregateResource` was called for because the `KubernetesCluster` specified an
+`AggregateResourceClass` as its `classRef`:
+
+```yaml
+apiVersion: compute.crossplane.io/v1beta1
+kind: KubernetesCluster
+metadata:
+  namespace: default
+  name: coolcluster
+spec:
+  resourceRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResource
+    name: default-coolcluster-g3bf7
+    uid: 2d45b85e-0cdd-11ea-8d71-362b9e155667
+  classRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResourceClass
+    name: cool-gke-cluster
+    uid: eee7e0a4-0ce2-11ea-8d71-362b9e155667
+```
+
+The creation of the above `KubernetesCluster` claim triggers:
+
+1. The dynamic provisioning of the `AggregateResource`, which aggregates all
+   managed resources labelled with the claim's namespace, name, and UID.
+1. The dynamic provisioning of a managed resource for each resource class that
+   aggregates to the `AggregateResourceClass` by matching its `aggregationRule`
+   labels. These dynamically provisioned managed resources are labelled with the
+   claim's namespace, name, and UID and thus aggregate to the dynamically
+   provisioned `AggregateResource`.
+
+The `AggregateResource` binds to the `KubernetesCluster` claim, while the
+dynamically provisioned managed resources bind in turn to the
+`AggregateResource`, thus creating a `resource claim -> aggregate resource ->
+managed resource(s)` binding relationship. The below `GKECluster` is
+_aggregated_. It binds to an `AggregateResource` (per its `aggregateRef` and
+`bindingPhase`) rather than to a resource claim:
+
+```yaml
+apiVersion: compute.gcp.crossplane.io/v1beta1
+kind: GKECluster
+metadata:
+  name: default-coolcluster-ffm8f
+  labels:
+    resourceclass.crossplane.io/aggregate-to: cool-gke-cluster
+spec:
+  forProvider:
+    # Omitted for brevity.
+  classRef:
+    apiVersion: compute.gcp.crossplane.io/v1beta1
+    kind: GKEClusterClass
+    name: cool-gke-cluster
+    uid: 64eeaf54-0cdd-11ea-8d71-362b9e155667
+  aggregateRef:
```

Comment:
Right, this would require every managed resource kind to support the new `aggregateRef` field, as well as setters and getters for that field at the code level. This would come "for free" with a new version of crossplane-runtime because all managed resource kinds embed a `ResourceSpec` from crossplane-runtime that handles this.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility

I'm pretty sure the addition of this new field would be handled by older controllers, which would simply ignore it.